File: builtin-templates/agent/automations.js
// automations.js — the Automations page (D83): the agents that work without
// anyone typing — schedules and watchers here; chat channels (auto-channels.js)
// and triggers register their own cards and details (registerKind). Each automation shows what it does, when,
// whose it is and how its last run went; open it for its runs (unread first
// to your eye), its settings, and — for a thread — "start afresh".
import { html, nothing, repeat } from '/vendor/lit-all.min.js';
import { selfApi as api, jbody } from '/vendor/bx-kit.js';

const KINDS = new Map();

/**
 * registerKind adds a kind of automation to the page.
 * @param kind  the API's kind (GET /automations items[].kind)
 * @param spec  {label, order, empty?: text when there are none,
 *   card?(page, item), head?(item, page) — the detail's header actions,
 *   detail?(item, page) — above its runs, runsLabel?: what its runs are called,
 *   open?(item id, page) — load what the detail shows (on open and on every
 *   refresh), create?: {label, start(page)} — a header button (start may set
 *   page.custom, a page of its own: custom(page) → template), load?(page) —
 *   extra data on every refresh, listExtra?(page) — below its cards}
 */
export function registerKind(kind, spec) { KINDS.set(kind, spec); }

const CADENCES = [
  ['0 9 * * *', 'every day at 9:00'],
  ['0 9 * * 1-5', 'every weekday at 9:00'],
  ['0 9 * * 1', 'every Monday at 9:00'],
  ['@every 1h', 'every hour'],
  ['@every 15m', 'every 15 minutes'],
];
const MODES = { isolated: 'a new run each time', persistent: 'one ongoing thread', conversation: 'into a conversation' };

export class AutoPage {
  /** @param on {change(), select(runId), me() → GET /me, route(kind, id) — what is open, for the address} */
  constructor(on) {
    this.on = on;
    this.items = [];
    this.summary = { count: 0, unread: 0, failing: 0 };
    this.open = null;   // {kind, id}
    this.runs = [];
    this.next = '';
    this.form = null;   // the schedule being created or edited
    this.custom = null; // a kind's own page (its form): custom(page) → template
    this.err = '';
  }

  changed() { this.on.change?.(); }

  async loadSummary() {
    try { this.summary = await api('/automations?summary=1'); } catch { /* keep */ }
    this.changed();
  }

  async load() {
    try { this.items = (await api('/automations')).items || []; this.err = ''; } catch (e) { this.err = e.message; }
    await Promise.all([...KINDS.values()].map((s) => s.load?.(this)));
    if (this.open) await Promise.all([this.loadRuns(), KINDS.get(this.open.kind)?.open?.(this.open.id, this)]);
    this.changed();
  }

  item() { return this.open && this.items.find((i) => i.kind === this.open.kind && i.id === this.open.id); }

  async show(kind, id) {
    this.open = kind ? { kind, id } : null;
    this.form = null;
    this.custom = null;
    this.on.route?.(kind, id); // a reload (or a copied link) comes back here
    this.runs = [];
    this.next = '';
    if (this.open) {
      await Promise.all([this.loadRuns(), KINDS.get(kind)?.open?.(id, this)]);
      api(`/automations/${kind}/${id}/read`, { method: 'POST' }).then(() => this.loadSummary()).catch(() => {});
    }
    this.changed();
  }

  async loadRuns(more = false) {
    const { kind, id } = this.open;
    const q = more && this.next ? `?cursor=${encodeURIComponent(this.next)}` : '';
    try {
      const p = await api(`/automations/${kind}/${id}/runs${q}`);
      this.runs = more ? [...this.runs, ...(p.items || [])] : p.items || [];
      this.next = p.next || '';
    } catch (e) { this.err = e.message; }
    this.changed();
  }

  newSchedule(watcher = false) {
    this.open = null;
    this.form = { name: '', cron: CADENCES[0][0], goal: '', mode: 'isolated', toolset: 'private', visibility: 'private', watcher };
    this.changed();
  }

  editSchedule(it) {
    const c = it.config || {};
    this.form = { id: it.id, name: it.name, cron: c.cron, goal: c.goal, system: c.system || '', mode: it.mode || 'isolated',
      toolset: c.toolset || 'private', visibility: it.visibility, watcher: it.kind === 'watcher', targetRun: it.targetRun };
    this.changed();
  }

  async save() {
    const f = this.form;
    if (!f.cron.trim() || !f.goal.trim()) { this.err = 'a cadence and what to do are needed'; this.changed(); return; }
    const body = { name: f.name.trim(), cron: f.cron.trim(), goal: f.goal.trim(), mode: f.watcher ? '' : f.mode,
      visibility: f.visibility, watcher: f.watcher, toolset: f.toolset, targetRun: f.targetRun || 0 };
    try {
      const s = f.id ? await api(`/schedules/${f.id}`, jbody(body, 'PUT')) : await api('/schedules', jbody(body, 'POST'));
      this.form = null;
      this.err = '';
      await this.load();
      await this.show(s.watcher ? 'watcher' : 'schedule', s.id);
    } catch (e) { this.err = e.message; this.changed(); }
  }

  async act(fn) {
    this.err = '';
    try { await fn(); await this.load(); } catch (e) { this.err = e.message; this.changed(); }
  }
  toggle(it) { return this.act(() => api(`/schedules/${it.id}`, jbody({ enabled: !it.enabled }, 'PUT'))); }
  runNow(it) { return this.act(() => api(`/schedules/${it.id}/trigger`, { method: 'POST' })); }
  reset(it) { return this.act(() => api(`/automations/${it.kind}/${it.id}/reset`, { method: 'POST' })); }
  async del(it) {
    if (!confirm(`Delete "${it.name}"? Its runs stay.`)) return;
    await this.act(() => api(`/schedules/${it.id}`, { method: 'DELETE' }));
    this.open = null;
    this.changed();
  }
}

registerKind('schedule', { label: 'Schedules', order: 1 });
registerKind('watcher', { label: 'Watchers', order: 2 });

export const ago = (sec) => {
  if (!sec) return '';
  const s = Math.max(0, Date.now() / 1000 - sec);
  return s < 90 ? 'just now' : s < 5400 ? `${Math.round(s / 60)} min ago` : s < 129600 ? `${Math.round(s / 3600)} h ago` : `${Math.round(s / 86400)} d ago`;
};
const cadence = (cron) => (CADENCES.find(([c]) => c === cron) || [cron, cron])[1];

// autoPageTpl is the page (drawn into the timeline).
export function autoPageTpl(p) {
  if (p.custom) return p.custom(p);
  if (p.form) return formTpl(p);
  const it = p.item();
  if (p.open && it) return detailTpl(p, it);
  const groups = [...KINDS.entries()].sort((a, b) => a[1].order - b[1].order)
    .map(([kind, spec]) => ({ kind, spec, items: p.items.filter((i) => i.kind === kind) }));
  return html`<div class="autos-page">
    <div class="ahd"><h3>Automations</h3><span class="muted small">agents that work without anyone typing</span>
      <span style="flex:1"></span>
      <button class="btn btnsm" @click=${() => p.newSchedule(false)}>New schedule</button>
      <button class="btn ghost btnsm" @click=${() => p.newSchedule(true)}>New watcher</button>
      ${[...KINDS.values()].filter((s) => s.create).map((s) => html`<button class="btn ghost btnsm" @click=${() => s.create.start(p)}>${s.create.label}</button>`)}</div>
    ${p.err ? html`<div class="err">${p.err}</div>` : nothing}
    ${groups.map((g) => html`<h5>${g.spec.label}</h5>
      ${g.items.length ? repeat(g.items, (i) => i.kind + i.id, (i) => (g.spec.card || cardTpl)(p, i))
        : html`<div class="muted small empty-line">${g.spec.empty || 'none yet'}</div>`}
      ${g.spec.listExtra ? g.spec.listExtra(p) : nothing}`)}
  </div>`;
}

function cardTpl(p, it) {
  const mine = it.access === 'owner';
  const failed = (it.lastStatus || '').startsWith('error');
  return html`<div class="acard2 ${it.enabled ? '' : 'off'}" data-auto=${it.kind + ':' + it.id} @click=${() => p.show(it.kind, it.id)}>
    <div class="ah"><span class="nm">${it.name}</span>
      ${it.unread ? html`<span class="badge unread">${it.unread} new</span>` : nothing}
      ${failed ? html`<span class="badge error" title=${it.lastStatus}>failed</span>` : nothing}
      ${it.enabled ? nothing : html`<span class="badge">off</span>`}
      <span style="flex:1"></span>
      ${it.access === 'oversee' ? html`<span class="muted small">${it.owner}'s</span>` : !mine && it.owner ? html`<span class="muted small">by ${it.owner}</span>` : nothing}
      ${mine ? html`<button class="btn ghost btnsm" @click=${(e) => { e.stopPropagation(); p.runNow(it); }}>Run now</button>` : nothing}
    </div>
    <div class="as muted small">${it.config ? cadence(it.config.cron) : it.summary}${it.config ? ' · ' + (it.config.goal || '').slice(0, 140) : ''}</div>
    <div class="as muted small">${it.kind === 'schedule' ? MODES[it.mode] || '' : 'keeps only the rounds where something changed'}
      ${it.lastRunAt ? ` · last ${ago(it.lastRunAt)}` : ''}${it.runs ? ` · ${it.runs} run${it.runs === 1 ? '' : 's'}` : ''}</div>
  </div>`;
}

function detailTpl(p, it) {
  const spec = KINDS.get(it.kind) || {};
  return html`<div class="autos-page">
    <div class="ahd"><a class="crumb" @click=${() => p.show(null)}>Automations</a> › <b>${it.name}</b>
      <span style="flex:1"></span>${(spec.head || scheduleHead)(it, p)}</div>
    ${p.err ? html`<div class="err">${p.err}</div>` : nothing}
    ${(spec.detail || scheduleDetail)(it, p)}
    <h5>${spec.runsLabel || 'Runs'}</h5>
    ${p.runs.length ? repeat(p.runs, (r) => r.id, (r) => html`<div class="run ${r.unread ? 'unread' : ''}" data-id=${r.id} @click=${() => p.on.select(r.id)}>
        <div class="t">${r.title || 'run ' + r.id}</div>
        <span class="gl">${new Date(r.activityMs).toLocaleString()}</span>
        ${r.status === 'error' ? html`<span class="gl err">!</span>` : r.status === 'running' ? html`<span class="spin"></span>` : nothing}
      </div>`) : html`<div class="muted small empty-line">none yet</div>`}
    ${p.next ? html`<button class="btn ghost btnsm" @click=${() => p.loadRuns(true)}>more</button>` : nothing}
  </div>`;
}

function scheduleHead(it, p) {
  const mine = it.access === 'owner';
  const thread = it.kind === 'watcher' || it.mode === 'persistent';
  return html`${mine ? html`<button class="btn ghost btnsm" @click=${() => p.runNow(it)}>Run now</button>` : nothing}
    ${mine || it.access === 'oversee' ? html`<label class="chk small"><input type="checkbox" .checked=${it.enabled} @change=${() => p.toggle(it)}> on</label>` : nothing}
    ${mine && it.config ? html`<button class="btn ghost btnsm" @click=${() => p.editSchedule(it)}>Edit</button>` : nothing}
    ${mine && thread ? html`<button class="btn ghost btnsm" title="its next run starts a new conversation; the old ones stay" @click=${() => p.reset(it)}>Start afresh</button>` : nothing}
    ${mine || it.access === 'oversee' ? html`<button class="btn rm btnsm" @click=${() => p.del(it)}>Delete</button>` : nothing}`;
}

function scheduleDetail(it, p) {
  return html`<div class="muted small">${it.config ? html`${cadence(it.config.cron)} · ${it.kind === 'schedule' ? MODES[it.mode] : 'a watcher'}
      ${it.mode === 'conversation' && it.targetRun ? html` · <a @click=${() => p.on.select(it.targetRun)}>reports to its conversation</a>` : nothing}`
      : it.summary}${it.lastStatus ? ` · last run: ${it.lastStatus}` : ''}</div>
    ${it.config && it.config.goal ? html`<div class="agoal">${it.config.goal}</div>` : nothing}`;
}

function formTpl(p) {
  const f = p.form;
  const set = (k) => (e) => { f[k] = e.target.type === 'checkbox' ? e.target.checked : e.target.value; p.changed(); };
  const preset = CADENCES.some(([c]) => c === f.cron);
  return html`<div class="autos-page">
    <div class="ahd"><a class="crumb" @click=${() => { p.form = null; p.changed(); }}>Automations</a> ›
      <b>${f.id ? 'Edit ' + (f.watcher ? 'watcher' : 'schedule') : f.watcher ? 'New watcher' : 'New schedule'}</b></div>
    ${p.err ? html`<div class="err">${p.err}</div>` : nothing}
    <div class="field"><label>Name</label><input .value=${f.name} @input=${set('name')} placeholder="Morning digest"></div>
    <div class="field"><label>When</label>
      <select @change=${(e) => { if (e.target.value) { f.cron = e.target.value; p.changed(); } }}>
        ${CADENCES.map(([c, l]) => html`<option value=${c} ?selected=${f.cron === c}>${l}</option>`)}
        <option value="" ?selected=${!preset}>custom…</option></select>
      ${preset ? nothing : html`<input class="mono" .value=${f.cron} @input=${set('cron')} placeholder="0 9 * * * or @every 30m">`}</div>
    <div class="field"><label>${f.watcher ? 'What to watch' : 'What to do'}</label>
      <textarea rows="3" .value=${f.goal} @input=${set('goal')}></textarea></div>
    ${f.watcher ? nothing : html`<div class="field"><label>Where each run goes</label>
      <select @change=${set('mode')}>
        <option value="isolated" ?selected=${f.mode === 'isolated'}>a new run each time (a report)</option>
        <option value="persistent" ?selected=${f.mode === 'persistent'}>one ongoing thread that builds on the last</option>
        ${f.mode === 'conversation' ? html`<option value="conversation" selected>into its conversation</option>` : nothing}
      </select></div>`}
    <div class="row2">
      <div class="field"><label>Tool mode</label><select @change=${set('toolset')} ?disabled=${!!f.id}>
        <option value="private" ?selected=${f.toolset !== 'web'}>internal systems, no web</option>
        <option value="web" ?selected=${f.toolset === 'web'}>web, no internal systems</option></select></div>
      <div class="field"><label>Who can see its runs</label><select @change=${set('visibility')}>
        <option value="private" ?selected=${f.visibility !== 'team'}>only you</option>
        <option value="team" ?selected=${f.visibility === 'team'}>everyone who can open this agent</option></select></div>
    </div>
    <div><button class="btn" @click=${() => p.save()}>${f.id ? 'Save' : 'Create'}</button></div>
  </div>`;
}

// sideEntryTpl is the sidebar's entry: the page, with what is new there.
export function sideEntryTpl(p, on, open) {
  const n = (p.summary.unread || 0) + (p.summary.attention || 0);
  return html`<div class="autos-entry ${on ? 'on' : ''}" @click=${open}>
    <span>Automations</span>
    ${n ? html`<span class="badge unread">${n}</span>` : nothing}
    ${p.summary.failing ? html`<span class="badge error" title="an automation's last run failed">!</span>` : nothing}
  </div>`;
}

File: builtin-templates/agent/_backend/automations.go
// automations.go — the non-UI agents in one list (D83): schedules and
// watchers now; channels and triggers register the same way when they land.
// Each automation owns runs (origin, origin_id); the Automations page shows
// an automation's runs and what is unread in them.
package main

import (
	"fmt"
	"net/http"
	"sort"
	"strconv"
	"strings"
	"time"

	xbin "github.com/xbin-dev/xbin/sdk"
)

// AutomationItem is one automation as the page lists it.
type AutomationItem struct {
	Kind       string `json:"kind"`
	ID         int64  `json:"id"`
	Name       string `json:"name"`
	Owner      string `json:"owner"`
	Visibility string `json:"visibility"`
	Access     string `json:"access"` // owner | viewer | oversee (a manager's view of someone else's)
	Enabled    bool   `json:"enabled"`
	Summary    string `json:"summary"` // one line: when and what
	Mode       string `json:"mode,omitempty"`
	TargetRun  int64  `json:"targetRun,omitempty"`
	CurrentRun int64  `json:"currentRun,omitempty"` // a thread's run now
	LastRunID  int64  `json:"lastRunId,omitempty"`
	LastRunAt  int64  `json:"lastRunAt,omitempty"`
	LastStatus string `json:"lastStatus,omitempty"`
	Runs       int    `json:"runs"`
	Unread     int    `json:"unread"`
	Attention  int    `json:"attention,omitempty"` // things waiting on the caller: a channel to claim, pairing requests, failed replies
	Config     any    `json:"config,omitempty"`    // the kind's own fields (the detail view)
}

// automationKind is one kind of automation. Origin is the run origin its runs
// carry.
type automationKind struct {
	Kind, Origin string
	List         func(w who) []AutomationItem
	Reset        func(w who, id int64) error // nil: nothing to reset
}

var automationKinds = map[string]automationKind{}

func registerAutomationKind(k automationKind) { automationKinds[k.Kind] = k }

func init() {
	for _, kind := range []string{"schedule", "watcher"} {
		k := kind
		registerAutomationKind(automationKind{Kind: k, Origin: k,
			List: func(w who) []AutomationItem { return scheduleItems(w, k == "watcher") },
			Reset: func(w who, id int64) error {
				s, err := agent.db.getSchedule(id)
				if err != nil || s.Watcher != (k == "watcher") {
					return fmt.Errorf("no such %s", k)
				}
				if s.access(w) < lvOwner {
					return fmt.Errorf("only its owner can start it afresh")
				}
				key := "sched:"
				if s.Watcher {
					key = "watch:"
					agent.db.setScheduleRun(id, 0)
				}
				agent.db.resetSession(key + strconv.FormatInt(id, 10))
				return nil
			}})
	}
}

func scheduleItems(w who, watchers bool) []AutomationItem {
	list, _ := agent.db.listSchedules()
	var out []AutomationItem
	for _, s := range list {
		if s.Watcher != watchers {
			continue
		}
		access := s.access(w)
		it := AutomationItem{Kind: "schedule", ID: s.ID, Name: orStr(s.Name, clip(s.Goal, 60)), Owner: s.Owner, Visibility: s.Visibility,
			Enabled: s.Enabled, Mode: orStr(s.Mode, modeIsolated), TargetRun: s.TargetRun, LastRunID: s.LastRunID,
			LastRunAt: s.LastRun, LastStatus: s.LastStatus,
			Summary: s.Cron + " · " + clip(s.Goal, 120),
			Config:  map[string]any{"cron": s.Cron, "goal": s.Goal, "system": s.System, "toolset": s.Toolset, "createdByRun": s.CreatedByRun}}
		if watchers {
			it.Kind, it.Mode, it.CurrentRun = "watcher", "", s.RunID
		} else if s.Mode == modePersistent {
			it.CurrentRun, _ = agent.db.sessionRun("sched:" + strconv.FormatInt(s.ID, 10))
		}
		switch {
		case access >= lvOwner:
			it.Access = "owner"
		case access >= lvViewer:
			it.Access = "viewer"
		case w.manager() && w.viewedBy == "":
			// oversight: it exists, it runs, whose it is — not what it does
			it.Access, it.Summary, it.Config = "oversee", s.Cron, nil
		default:
			continue
		}
		out = append(out, it)
	}
	return out
}

// automationCounts is, per (origin, origin_id), how many runs the caller may
// see and how many are unread for them.
func automationCounts(w who) map[string][2]int {
	where, args := aclWhere(w)
	epoch := agent.convEpoch()
	q := `SELECT r.origin, r.origin_id, count(*), SUM(CASE WHEN r.activity_ms > max(COALESCE(us.read_ms,0), ?) THEN 1 ELSE 0 END)
		FROM runs r LEFT JOIN run_user_state us ON us.run_id=r.id AND us.user=?
		WHERE r.parent_id=0 AND r.origin NOT IN ('', 'chat', 'api') AND ` + where + ` GROUP BY r.origin, r.origin_id`
	rows, err := agent.db.q.Query(q, append([]any{epoch, w.user}, args...)...)
	out := map[string][2]int{}
	if err != nil {
		return out
	}
	defer rows.Close()
	for rows.Next() {
		var origin string
		var id int64
		var n, unread int
		if rows.Scan(&origin, &id, &n, &unread) == nil {
			if w.kind != whoUser {
				unread = 0
			}
			out[origin+":"+strconv.FormatInt(id, 10)] = [2]int{n, unread}
		}
	}
	return out
}

func allAutomations(w who) []AutomationItem {
	counts := automationCounts(w)
	var out []AutomationItem
	for _, k := range automationKinds {
		for _, it := range k.List(w) {
			c := counts[k.Origin+":"+strconv.FormatInt(it.ID, 10)]
			it.Runs, it.Unread = c[0], c[1]
			out = append(out, it)
		}
	}
	sort.Slice(out, func(i, j int) bool {
		if out[i].Kind != out[j].Kind {
			return out[i].Kind < out[j].Kind
		}
		return out[i].ID < out[j].ID
	})
	return out
}

// handleAutomations lists the automations the caller may see (and, for a
// manager, every automation's existence). ?summary=1 is just the counts the
// sidebar badge needs.
func handleAutomations(w http.ResponseWriter, r *http.Request) {
	c := callerOf(r)
	items := allAutomations(c)
	if r.URL.Query().Get("summary") == "1" {
		unread, failing, attention := 0, 0, 0
		for _, it := range items {
			if it.Access == "owner" || it.Access == "viewer" {
				unread += it.Unread
				if strings.HasPrefix(it.LastStatus, "error") {
					failing++
				}
			}
			if it.Access == "owner" || it.Access == "claim" {
				attention += it.Attention
			}
		}
		xbin.WriteJSON(w, 200, map[string]any{"count": len(items), "unread": unread, "failing": failing, "attention": attention})
		return
	}
	if items == nil {
		items = []AutomationItem{}
	}
	xbin.WriteJSON(w, 200, map[string]any{"items": items})
}

func automationFor(w http.ResponseWriter, r *http.Request) (*AutomationItem, automationKind, bool) {
	c := callerOf(r)
	k, ok := automationKinds[r.PathValue("kind")]
	id, _ := strconv.ParseInt(r.PathValue("aid"), 10, 64)
	if ok {
		counts := automationCounts(c)
		for _, it := range k.List(c) {
			if it.ID == id {
				cnt := counts[k.Origin+":"+strconv.FormatInt(id, 10)]
				it.Runs, it.Unread = cnt[0], cnt[1]
				return &it, k, true
			}
		}
	}
	xbin.WriteError(w, 404, "no such automation")
	return nil, k, false
}

func handleAutomation(w http.ResponseWriter, r *http.Request) {
	if it, _, ok := automationFor(w, r); ok {
		xbin.WriteJSON(w, 200, it)
	}
}

// handleAutomationRuns pages an automation's runs, newest activity first,
// as conversation rows (unread per caller).
//
//	GET /automations/{kind}/{aid}/runs?cursor=&limit=
func handleAutomationRuns(w http.ResponseWriter, r *http.Request) {
	it, k, ok := automationFor(w, r)
	if !ok {
		return
	}
	c := callerOf(r)
	limit, _ := strconv.Atoi(r.URL.Query().Get("limit"))
	if limit <= 0 || limit > 100 {
		limit = 30
	}
	where, args := aclWhere(c)
	q := `r WHERE r.parent_id=0 AND r.origin=? AND r.origin_id=? AND ` + where
	args = append([]any{k.Origin, it.ID}, args...)
	if cur := r.URL.Query().Get("cursor"); cur != "" {
		ms, id, ok := parseConvCursor(cur)
		if !ok {
			xbin.WriteError(w, 400, "bad cursor")
			return
		}
		q += ` AND (r.activity_ms<? OR (r.activity_ms=? AND r.id<?))`
		args = append(args, ms, ms, id)
	}
	runs, err := agent.db.queryRuns(q+` ORDER BY r.activity_ms DESC, r.id DESC LIMIT `+strconv.Itoa(limit+1), args...)
	if err != nil {
		xbin.WriteError(w, 500, err.Error())
		return
	}
	next := ""
	if len(runs) > limit {
		runs = runs[:limit]
		next = fmt.Sprintf("%d.%d", runs[len(runs)-1].ActivityMs, runs[len(runs)-1].ID)
	}
	var ids []int64
	for _, x := range runs {
		ids = append(ids, x.ID)
	}
	states := agent.db.userStates(c.user, ids)
	items := []map[string]any{}
	for _, x := range runs {
		items = append(items, agent.convItem(x, c, states[x.ID]))
	}
	xbin.WriteJSON(w, 200, map[string]any{"items": items, "next": next})
}

// handleAutomationRead marks every run of an automation read for the caller.
func handleAutomationRead(w http.ResponseWriter, r *http.Request) {
	it, k, ok := automationFor(w, r)
	if !ok {
		return
	}
	c := callerOf(r)
	if c.kind == whoUser {
		ms := time.Now().UnixMilli()
		where, args := aclWhere(c)
		_, _ = agent.db.q.Exec(`INSERT INTO run_user_state (run_id, user, read_ms)
			SELECT r.id, ?, ? FROM runs r WHERE r.parent_id=0 AND r.origin=? AND r.origin_id=? AND `+where+`
			ON CONFLICT(run_id, user) DO UPDATE SET read_ms=excluded.read_ms`,
			append([]any{c.user, ms, k.Origin, it.ID}, args...)...)
	}
	xbin.WriteJSON(w, 200, map[string]string{"ok": "true"})
}

// handleAutomationReset starts a thread (a persistent schedule, a watcher)
// afresh: its next firing opens a new run; the old ones stay listed.
func handleAutomationReset(w http.ResponseWriter, r *http.Request) {
	it, k, ok := automationFor(w, r)
	if !ok {
		return
	}
	if k.Reset == nil {
		xbin.WriteError(w, 400, "nothing to reset")
		return
	}
	if err := k.Reset(callerOf(r), it.ID); err != nil {
		xbin.WriteError(w, 403, err.Error())
		return
	}
	xbin.WriteJSON(w, 200, map[string]string{"ok": "true"})
}
